### ISSUE TITLE
coqchk times out on some files that coqc accepts ~instantly

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#12200]
#### Description of the problem

Consider the code
```coq
Fixpoint fact (n : nat)
  := match n with
     | 0 => 1
     | S n => (S n) * fact n
     end.

Goal id 0 = 0.
  Strategy expand [id].
  assert (id (fact 100) = fact 100) by abstract reflexivity.
  Opaque id.
  let fact' := (eval cbv in fact) in assert (id (fact' (id 100)) = fact (id 100)) by abstract reflexivity.
  reflexivity.
Qed.
```
coqc will accept this file ~instantly, but coqchk spins for as long as I was willing to wait (projected to take more than the lifetime of the universe).  This is because `Strategy` adds its data to the libobject stack, but this data is not interwoven with proofs.  (In fact, `coqchk` already times out even if we remove the `let fact' := (eval cbv in fact) in assert (id (fact' (id 100)) = fact (id 100)) by abstract reflexivity.` line.)  I believe transparency information should instead be somehow recorded in the proof term in the relevant cast nodes.

#### Coq Version

8.11.1

**Comment by andres-erbsen** [comment]
TIL coqchk supports strategies at all. Is there anything like documentation where one could possibly learn about this?

**Comment by ppedrot** [comment]
`coqchk` only understands strategies in a very limited way. First they need to be part of the kernel, i.e.:
- standard lazy
- call-by-name
- VM call-by-value
- native call-by-value.
Furthermore there are restrictions on how it behaves. Cast nodes have a clear tendency to disappear during reduction, so it's hard to understand what is going on when sending a strategy-annotated term to the kernel.

**Comment by SkySkimmer** [comment]
coqchk doesn't use VM or native even if there's a vmcast/nativecast afaik

**Comment by ppedrot** [comment]
@SkySkimmer you're right, I don't know what I was thinking about when I wrote this. My comment applies to the kernel, not the checker, which is even more restrictive (although it uses the same underlying mechanism). VM and native could be added to the checker, but they are deemed too unsafe to be trusted.
